### ISSUE TITLE
Unrecoverable crash on Goerli - data corruption

### ISSUE THREAD

**Issue by gus4rs** [status-im/nim-beacon-chain#4063]
Version:  ```v22.8.2```

The process crashed with the following log:

```
 WRN 2022-09-04 21:01:11.000+01:00 Cannot construct EpochRef for duties - report bug topics="beacval" head=4ef52a2d:3817772 slot=3817806
Sep 04 21:01:11 home nimbus_beacon_node[9237]: NOT 2022-09-04 21:01:11.473+01:00 Received invalid sequence of blocks        topics="syncman" request=3817772:27@4 peer=16U*EiXVaj direction=forward blocks_count=24 blocks_map=xxx.xxxxx.xxx>
Sep 04 21:01:12 home nimbus_beacon_node[9237]: INF 2022-09-04 21:01:12.000+01:00 Slot start                                 topics="beacnde" slot=3817806 epoch=119306 sync="--h--m (100.00%) 0.0000slots/s (RRRRRRRQRR:3817771)" peers=25 h>
Sep 04 21:01:12 home nimbus_beacon_node[9237]: WRN 2022-09-04 21:01:12.005+01:00 Fork choice selected unknown head, trying to sync topics="attpool" root=4ef52a2d
Sep 04 21:01:12 home nimbus_beacon_node[9237]: WRN 2022-09-04 21:01:12.005+01:00 Head selection failed, using previous head head=4ef52a2d:3817772 wallSlot=3817806
Sep 04 21:01:12 home nimbus_beacon_node[9237]: NOT 2022-09-04 21:01:12.727+01:00 Received invalid sequence of blocks        topics="syncman" request=3817772:27@5 peer=16U*JzqT9H direction=forward blocks_count=24 blocks_map=xxx.xxxxx.xxx>
Sep 04 21:01:13 home nimbus_beacon_node[9237]: NOT 2022-09-04 21:01:13.526+01:00 Received invalid sequence of blocks        topics="syncman" request=3817772:27@6 peer=16U*xYvuqn direction=forward blocks_count=24 blocks_map=xxx.xxxxx.xxx>
Sep 04 21:01:14 home nimbus_beacon_node[9237]: NOT 2022-09-04 21:01:14.810+01:00 Received invalid sequence of blocks        topics="syncman" request=3817772:27@7 peer=16U*pb6uMo direction=forward blocks_count=24 blocks_map=xxx.xxxxx.xxx>
Sep 04 21:01:16 home nimbus_beacon_node[9237]: WRN 2022-09-04 21:01:16.005+01:00 Fork choice selected unknown head, trying to sync topics="attpool" root=4ef52a2d
Sep 04 21:01:16 home nimbus_beacon_node[9237]: WRN 2022-09-04 21:01:16.005+01:00 Head selection failed, using previous head head=4ef52a2d:3817772 wallSlot=3817806
Sep 04 21:01:16 home nimbus_beacon_node[9237]: WRN 2022-09-04 21:01:16.005+01:00 Cannot construct EpochRef for attestation head, report bug topics="beacval" attestationHead=4ef52a2d:3817772@3817806 slot=3817806
Sep 04 21:01:20 home nimbus_beacon_node[9237]: WRN 2022-09-04 21:01:20.000+01:00 Cannot construct EpochRef for head, report bug topics="beacval" head=4ef52a2d:3817772 slot=3817806
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-libp2p/libp2p/stream/bufferstream.nim(460) main
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-libp2p/libp2p/stream/bufferstream.nim(453) NimMain
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(2155) main
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(2023) handleStartUpCmd
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1861) doRunBeaconNode
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1663) start
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1606) run
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-chronos/chronos/asyncloop.nim(285) poll
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-chronos/chronos/asyncfutures2.nim(365) futureContinue
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1358) onSlotStart
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1162) onSlotEnd
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-chronos/chronos/asyncfutures2.nim(365) futureContinue
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1211) onSlotEnd
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-stew/stew/results.nim(756) expect
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-stew/stew/results.nim(348) raiseResultDefect
Sep 04 21:01:20 home nimbus_beacon_node[9237]: [[reraised from:
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-libp2p/libp2p/stream/bufferstream.nim(460) main
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-libp2p/libp2p/stream/bufferstream.nim(453) NimMain
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(2155) main
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(2023) handleStartUpCmd
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1861) doRunBeaconNode
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1663) start
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1606) run
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-chronos/chronos/asyncloop.nim(285) poll
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-chronos/chronos/asyncfutures2.nim(365) futureContinue
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1358) onSlotStart
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1162) onSlotEnd
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-chronos/chronos/asyncfutures2.nim(387) futureContinue
Sep 04 21:01:20 home nimbus_beacon_node[9237]: ]]
Sep 04 21:01:20 home nimbus_beacon_node[9237]: [[reraised from:
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-libp2p/libp2p/stream/bufferstream.nim(460) main
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-libp2p/libp2p/stream/bufferstream.nim(453) NimMain
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(2155) main
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(2023) handleStartUpCmd
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1861) doRunBeaconNode
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1663) start
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1606) run
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-chronos/chronos/asyncloop.nim(285) poll
Sep 04 21:01:20 home nimbus_beacon_node[9237]: /home/bitcoin/nimbus/vendor/nim-chronos/chronos/asyncfutures2.nim(387) futureContinue
```

When trying to restart it, it crashes again:

```
Sep 04 21:18:25 home nimbus_beacon_node[556177]: NOT 2022-09-04 21:18:25.770+01:00 Starting metrics HTTP server               topics="beacnde" url=http://127.0.0.1:8009/metrics
Sep 04 21:18:25 home nimbus_beacon_node[556177]: INF 2022-09-04 21:18:25.908+01:00 Threadpool started                         topics="beacnde" numThreads=8
Sep 04 21:18:28 home nimbus_beacon_node[556177]: INF 2022-09-04 21:18:28.749+01:00 Loading block DAG from database            topics="beacnde" path=/home/data/ethereum/testnet/consensus/nimbus/db
Sep 04 21:18:29 home nimbus_beacon_node[556177]: /home/bitcoin/nimbus/vendor/nim-libp2p/libp2p/stream/bufferstream.nim(460) main
Sep 04 21:18:29 home nimbus_beacon_node[556177]: /home/bitcoin/nimbus/vendor/nim-libp2p/libp2p/stream/bufferstream.nim(453) NimMain
Sep 04 21:18:29 home nimbus_beacon_node[556177]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(2155) main
Sep 04 21:18:29 home nimbus_beacon_node[556177]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(2023) handleStartUpCmd
Sep 04 21:18:29 home nimbus_beacon_node[556177]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(1841) doRunBeaconNode
Sep 04 21:18:29 home nimbus_beacon_node[556177]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(625) init
Sep 04 21:18:29 home nimbus_beacon_node[556177]: /home/bitcoin/nimbus/beacon_chain/nimbus_beacon_node.nim(168) loadChainDag
Sep 04 21:18:29 home nimbus_beacon_node[556177]: /home/bitcoin/nimbus/beacon_chain/consensus_object_pools/blockchain_dag.nim(795) init
Sep 04 21:18:29 home nimbus_beacon_node[556177]: /home/bitcoin/nimbus/vendor/nim-stew/stew/results.nim(756) expect
Sep 04 21:18:29 home nimbus_beacon_node[556177]: /home/bitcoin/nimbus/vendor/nim-stew/stew/results.nim(348) raiseResultDefect
Sep 04 21:18:29 home nimbus_beacon_node[556177]: Error: unhandled exception: Head root in database, corrupt? [ResultDefect]
```

**Comment by AntoBrajkovic** [comment]
I have the same data corruption issue on mainet 

/home/user/nimbus-eth2/vendor/nim-testutils/testutils/moduletests.nim(21) moduletests
/home/user/nimbus-eth2/beacon_chain/nimbus_beacon_node.nim(2163) main
/home/user/nimbus-eth2/beacon_chain/nimbus_beacon_node.nim(2031) handleStartUpCmd
/home/user/nimbus-eth2/beacon_chain/nimbus_beacon_node.nim(1848) doRunBeaconNode
/home/user/nimbus-eth2/beacon_chain/nimbus_beacon_node.nim(625) init


/home/user/nimbus-eth2/beacon_chain/nimbus_beacon_node.nim(168) loadChainDag
/home/user/nimbus-eth2/beacon_chain/consensus_object_pools/blockchain_dag.nim(795) init
/home/user/nimbus-eth2/vendor/nim-stew/stew/results.nim(756) expect
/home/user/nimbus-eth2/vendor/nim-stew/stew/results.nim(348) raiseResultDefect
Error: unhandled exception: Head root in database, corrupt? [ResultDefect]

**Comment by tersec** [comment]
https://github.com/status-im/nimbus-eth2/pull/4192 should fix this

**Comment by zah** [comment]
The fix was shipped with Nimbus 22.10.0
